tests/inference/ebm_test.py

```diff
@@ -785,6 +786,134 @@ def test_call(self):
+
+    initial_state = tf.constant([0] * num_bits, tf.int8)
+    initial_energy = test_energy(tf.expand_dims(initial_state, 0))
+
+    one_step_wrapper = tf.function(actual_kernel.one_step)
+    next_state, _ = one_step_wrapper(initial_state, [])
+    next_energy = test_energy(tf.expand_dims(next_state, 0))
+
+    self.assertNotAllEqual(initial_state, next_state)
+    self.assertGreater(initial_energy, next_energy)
+
+
+class GibbsWithGradientsInferenceTest(tf.test.TestCase):
+  """Tests the GibbsWithGradientsInference class."""
+
+  def setUp(self):
+    """Initializes test objects."""
+    super().setUp()
+    self.tf_random_seed = 4
+    self.tfp_seed = tf.constant([3, 4], tf.int32)
+    self.close_rtol = 1e-2
+    self.zero_atol = 1e-5
+    self.not_zero_atol = 1e-1
+
+  def test_init(self):
+    """Confirms internal values are set correctly."""
+    bits = [0, 1, 3]
+    order = 2
+    expected_energy = models.KOBE(bits, order)
+    expected_num_expectation_samples = 14899
+    expected_num_burnin_samples = 32641
+    expected_name = "test_analytic_dist_name"
+    actual_layer = inference.GibbsWithGradientsInference(
+        expected_energy, expected_num_expectation_samples,
+        expected_num_burnin_samples, expected_name)
+
+    self.assertEqual(actual_layer.energy, expected_energy)
+    self.assertAllEqual(actual_layer.num_expectation_samples,
+                        expected_num_expectation_samples)
+    self.assertAllEqual(actual_layer.num_burnin_samples,
+                        expected_num_burnin_samples)
+    self.assertEqual(actual_layer.name, expected_name)
+
+  @test_util.eager_mode_toggle
+  def test_sample(self):
+    """Confirms that bitstrings are sampled as expected."""
+    # Set up energy function
+    num_bits = 5
+    num_layers = 3
+    num_independent_chains = 5
+    bits = random.sample(range(1000), num_bits)
+    units = random.sample(range(1, 100), num_layers)
+    activations = random.sample([
+        "elu", "exponential", "gelu", "hard_sigmoid", "linear", "relu", "selu",
+        "sigmoid", "softmax", "softplus", "softsign", "swish", "tanh"
+    ], num_layers)
+    expected_layer_list = []
+    for i in range(num_layers):
+      expected_layer_list.append(
+          tf.keras.layers.Dense(units[i], activation=activations[i]))
+    expected_layer_list.append(tf.keras.layers.Dense(1))
+    expected_layer_list.append(utils.Squeeze(-1))
+    actual_energy = models.BitstringEnergy(bits, expected_layer_list)
+    _ = actual_energy(tf.constant([[0] * num_bits], dtype=tf.int8))
+
+    # Sampler
+    num_expectation_samples = int(1e3)
+    num_burnin_samples = int(1e2)
+    actual_layer = inference.GibbsWithGradientsInference(
+        actual_energy, num_expectation_samples, num_burnin_samples)
+
+    sample_wrapper = tf.function(actual_layer.sample)
+    samples = sample_wrapper(num_expectation_samples)
+
+    # Check that entropy of samples is approximately entropy of distribution,
```

Comment:
would comparing log partition be even stronger since that's what MCMC is solving for in a sense? I guess you may have avoided this intentionally because of the bias of the current estimator? that's reasonable, but I think that updating this test in the future with unbiased log-partition estimator would make this stronger

Reply:
Added a histogram test